src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,284 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        } else if (isEmpty()) {
+            addFirstElement(value);
+        } else if (index == 0) {
+            addElementInBeginning(value);
+        } else if (index == size) {
+            addElementInEnd(value);
+        } else if (index < size && value != null) {
+            addElementInTheMiddle(index, value);
+        }
+        size++;
+    }
+
+    private void addElementInBeginning(T value) {
+        head.prev = new Node<>(null, value, head);
+        head = head.prev;
+    }
+
+    private void addElementInEnd(T value) {
+        tail.next = new Node<>(tail, value, null);
+        tail = tail.next;
+    }
+
+    private void addFirstElement(T value) {
+        tail = new Node<>(null, value, null);
+        head = tail;
+    }
+
+    private void addElementInTheMiddle(int index, T value) {
+        Node<T> current = head;
+        int counter = 0;
+        int step = 1;
+        String iterationVariant = FROM_START_TYPE_OF_ITERATION;
+        if (index > size / 2) {
+            current = tail;
+            counter = size - 1;
+            step = -1;
+            iterationVariant = FROM_END_TYPE_OF_ITERATION;
+        }
+        while (current != null) {
+            if (counter == index) {
+                link(current, value);
+            }
+            current = iterationVariant.equals(FROM_START_TYPE_OF_ITERATION)
+                    ? current.next : current.prev;
+            counter += step;
+        }
+    }
+
+    private void link(Node<T> current, T value) {
+        Node<T> prev = current.prev;
+        Node<T> newElem = new Node<>(prev, value, current);
+        prev.next = newElem;
+        current.prev = newElem;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T elem : list) {
+                add(elem);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException(INDEX_OF_BOUND_EXCEPTION_MESSAGE);
+        }
+        Node<T> current = head;
+        int counter = 0;
+        int step = 1;
+        String iterationVariant = FROM_START_TYPE_OF_ITERATION;
+        if (index > size / 2) {
+            current = tail;
+            counter = size - 1;
+            step = -1;
+            iterationVariant = FROM_END_TYPE_OF_ITERATION;
+        }
+        while (current != null) {
+            if (counter == index) {
+                return current.value;
+            }
+            current = iterationVariant.equals(FROM_START_TYPE_OF_ITERATION)
+                    ? current.next : current.prev;
+            counter += step;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
```

Comment:
```suggestion
        if (index < 0 || index >= size) {
```
and move it to a private method. it is used more than once

Reply:
moved code to method checkSize() and fixed diapason